### PR TITLE
readonly fields

### PR DESCRIPTION
**Pull request by KATT**
#202

### CODE REVIEW

**Comment by KATT** [comment]
Maybe I should reverse this and instead of having a blacklist of fields, I'd have a whitelist. Whatcha think?

**Comment by KATT** [comment]
`editable: []` array

**Comment by Zertz** [comment]
Thanks! I'll take a look at this

See also #185

**Comment by KATT** [comment]
Ah, my apps doesn't currently have multiple roles - either you can access the API or you can't. Could start with this and then make it accessible on a role-based level?

**Comment by KATT** [comment]
We might wanna add some more tests on my `deleteDeep` function. Wanted it to conform to Mongo's way of accessing nested properties so I had to write it myself.

Locally I did these tests when I wrote it:

```jsx
import _ from 'lodash';

function deleteDeep(item, chunks) {
    var firstChunk = chunks[0]
    var firstValue = item[firstChunk]

    if (chunks.length === 1) {
        delete item[firstChunk]
    } else if (_.isArray(firstValue)) {
        firstValue.forEach(function (prop) {
            deleteDeep(prop, chunks.slice(1))
        })
    } else if (_.isObject(firstValue)) {
        deleteDeep(firstValue, chunks.slice(1))
    }

    return item
}


describe('deleteDeep', () => {
    it('should delete not deep', () => {
        const input = {
            foo : 'bar',
            foo2: 2,
        }
        const chunks = ['foo'];
        const output = {
            foo2: 2
        };
        expect(deleteDeep(input, chunks)).to.deep.eq(output);
    });

    it('should delete deep objects', () => {
        const input = {
            foo: {
                foofoo: {
                    foofoofoo: 'bar'
                }
            },
            foo2: 2,
        }
        const chunks = ['foo', 'foofoo', 'foofoofoo'];
        const output = {
            foo2: 2,
            foo : {
                foofoo: {
                }
            },
        };
        expect(deleteDeep(input, chunks)).to.deep.eq(output);
    });

    it('should delete deep arrays', () => {
        const input = {
            foos: [{
                wow: [
                    {
                        ignored    : 'okay',
                        not_ignored: 'yay',
                    }
                ]
            }],
            foo2: 2,
        }
        const chunks = ['foos', 'wow', 'ignored'];
        const output = {
            foo2: 2,
            foos: [{
                wow: [
                    {
                        not_ignored: 'yay',
                    }
                ]
            }],
        };
        expect(deleteDeep(input, chunks)).to.deep.eq(output);
    });
});
```

**Comment by KATT** [comment]
@Zertz have you had time to look at this yet? I've started using it on a project now and it works :+1:

**Comment by Zertz** [comment]
Hey @KATT I'm in Portland for node interactive, hence the delay!

I'll try to have a closer look tonight or tomorrow.

At a glance it looks good, although I think private and protected read-only fields instead of just a read-only blacklist would be even better.

I know I have many fields that should be read-only for a group of users, but read/writer for others.

```
privateReadOnly
protectedReadOnly
```

Is there a lodash functionor some other module that we could use instead of deleteDeep?

File: lib/resource_filter.js [hunk 1]
```diff
@@ -142,13 +142,40 @@ module.exports = function (model, filteredKeys) {
...
     opts = _.defaults(opts || {}, {
-      access: 'public'
+      access: 'public',
+      blacklist: []
     })
 
     var filtered = filterItem(resource, getExcluded(opts.access, model.modelName))
 
+    filtered = filterBlacklistedKeys(filtered, opts.blacklist)
+
```
Comment (Zertz) [inline lib/resource_filter.js]: This should probably be inside resource filter so we don't get too much controller bloat.

**Comment by KATT** [comment]
> Is there a lodash functionor some other module that we could use instead of deleteDeep?

I would like there to be, but I haven't found one that conforms to the "Mongo way" of accessing nested properties. To be specific, arrays are the problem.

Let's say we have a document like this:

```jsx
// users
{
  name: 'Alex',
  issues: [
    name: '#203',
    last_mod: '2015-12-09T11:56:06.332Z'
  ]
}
```  

We want to make sure that `last_mod` is never changed by an API-request. 

There's a some libraries I've found that can help with getting/mutating deep properties, but you also need to pass the array index in all of them. Same for [lodash](https://lodash.com/docs#get) as for [object-path](https://github.com/mariocasciaro/object-path).

We don't want to have `readonly` to be `['issues[0].lastmod', 'issues[1].lastmod', .., 'issues[N].lastmod']`, we want it the to be `['issues.lastmod']`. That' why I ended up writing it myself.

Please do tell me if you can find it done somewhere. It should've been done before.

Otherwise, if you want, I could break it out into it's own mini project with more thorough test coverage.


> I know I have many fields that should be read-only for a group of users, but read/writer for others.

Can we not do this in a later update? For now `readonly` would work as global readonly. We can add more granular controls later

**Comment by Zertz** [comment]
Great work, thanks for looking into this!

Yes, I think it should be in a separate module.

`filterBlacklistedKeys`  should perhaps be renamed to `filterReadOnlyKeys` instead? Definitely moved in resource_filter though and called within `filterItem`.

My opinion is that we should be consistent with the current private/protected implementation and provide both privateReadOnly and protectedReadOnly.

**Comment by KATT** [comment]
Will probably do the refactoring sometime this week.. always some time and pain with doing the boilerplating for a new npm package.

**Comment by Zertz** [comment]
It's alright! I'm busy this week, but let me know if you need help with anything

**Comment by KATT** [comment]
I have no use of this anymore - feel free to pick the code up where I left off.

Thanks for your support!

**Comment by Zertz** [comment]
Thanks for putting time into this, it's definitely an area where the project needs some work and it's very helpful to see other people's take on this issue.

We could probably make this part of a larger reworking of the resource filter.

**Pull request closed without merging**